File: language_construct_performance/ruby/scripts/benchmark_utils.rb
require 'json'
require 'fileutils'

NUM_ENTRIES = (ENV['NUM_ENTRIES'] || '20000000').to_i

def measure_memory
  GC.start(full_mark: true, immediate_sweep: true)

  rss_kb = 0
  if File.exist?('/proc/self/status')
    # Linux (Docker container)
    File.read('/proc/self/status').each_line do |line|
      if line.start_with?('VmRSS:')
        rss_kb = line.split[1].to_i
        break
      end
    end
  else
    # Windows fallback
    mem = `powershell -Command "(Get-Process -Id #{Process.pid}).WorkingSet64"`.strip.to_i
    rss_kb = mem / 1024
  end

  rss_mb = (rss_kb / 1024.0).round
  {
    rss: rss_mb,
    heap_used: rss_mb  # Ruby doesn't expose heap breakdown like JS; use RSS as proxy
  }
end

def benchmark_stats(name, stat_group, stat_key, track_memory: false)
  mem_before = track_memory ? measure_memory : nil
  t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC, :millisecond)
  result = yield
  t1 = Process.clock_gettime(Process::CLOCK_MONOTONIC, :millisecond)
  mem_after = track_memory ? measure_memory : nil

  time_ms = t1 - t0
  stat_group[stat_key] = time_ms

  if track_memory
    memory_mb = mem_after[:rss] - mem_before[:rss]
    stat_group['memoryUsedMB'] = memory_mb
    puts "#{name} - Time: #{time_ms}ms, Memory Used (RSS): #{memory_mb} MB #{mem_after.to_json}"
  else
    puts "#{name}: #{time_ms}ms"
  end

  result
end

def save_stats(file_name, key, data)
  FileUtils.mkdir_p('data')
  stats_path = File.join('data', file_name)
  existing_stats = {
    'numEntries' => NUM_ENTRIES,
    'plain object' => {},
    'value object' => {},
    'value object minimal' => {}
  }

  if File.exist?(stats_path)
    begin
      existing_stats = JSON.parse(File.read(stats_path))
    rescue JSON::ParserError => e
      $stderr.puts "Error parsing existing stats in #{stats_path}: #{e.message}"
    end
  end

  existing_stats[key] = data
  existing_stats['numEntries'] = NUM_ENTRIES

  File.write(stats_path, JSON.pretty_generate(existing_stats))
  puts "\nSaved #{key} stats to data/#{file_name}"
end

File: language_construct_performance/nodejs/scripts/benchmark_utils.mjs
import { writeFileSync, readFileSync, mkdirSync, existsSync } from "fs";
import { join } from "path";
import { performance } from "perf_hooks";

export const NUM_ENTRIES = parseInt(process.env.NUM_ENTRIES || "20000000", 10);

export function measureMemory() {
  if (global.gc) global.gc(); // Force GC before taking memory snapshot
  const mem = process.memoryUsage();
  return {
    rss: Math.round(mem.rss / 1024 / 1024),
    heapTotal: Math.round(mem.heapTotal / 1024 / 1024),
    heapUsed: Math.round(mem.heapUsed / 1024 / 1024),
    external: Math.round(mem.external / 1024 / 1024),
  };
}

export function benchmarkStats(
  name,
  statGroup,
  statKey,
  fn,
  trackMemory = false
) {
  const memBefore = trackMemory ? measureMemory() : null;
  const t0 = performance.now();
  const result = fn();
  const t1 = performance.now();
  const memAfter = trackMemory ? measureMemory() : null;

  const timeMs = Math.round(t1 - t0);
  statGroup[statKey] = timeMs;

  let memoryMB;
  if (trackMemory) {
    memoryMB = memAfter.heapUsed - memBefore.heapUsed;
    statGroup.memoryUsedMB = memoryMB;
    console.log(`${name} - Time: ${timeMs}ms, Memory Used (Heap): ${memoryMB} MB ${JSON.stringify(memAfter)}`);
  } else {
    console.log(`${name}: ${timeMs}ms`);
  }

  return result;
}

export function saveStats(fileName, key, data) {
  if (!existsSync("data")) mkdirSync("data");
  const statsPath = join("data", fileName);
  let existingStats = { numEntries: NUM_ENTRIES, "plain object": {}, "value object": {}, "value object minimal": {} };
  if (existsSync(statsPath)) {
    try {
      existingStats = JSON.parse(readFileSync(statsPath, "utf-8"));
    } catch (e) {
      console.error(`Error parsing existing stats in ${statsPath}:`, e);
    }
  }

  existingStats[key] = data;
  existingStats.numEntries = NUM_ENTRIES;

  writeFileSync(statsPath, JSON.stringify(existingStats, null, 2));
  console.log(`\nSaved ${key} stats to data/${fileName}`);
}

export function forceGC() {
  if (global.gc) {
    global.gc();
  } else {
    console.warn("GC not exposed. Run with --expose-gc flag for accurate memory measurements.");
  }
}
